arraycontext/impl/pytato/compile.py

```diff
@@ -0,0 +1,283 @@
+        Before calling :attr:`~PytatoCompiledOperator.f`, it is compiled to a
+        :mod:`pytato` DAG that would apply :attr:`~PytatoCompiledOperator.f`
+        with *args* in a lazy-sense.
+        """
+
+        from arraycontext.container.traversal import (rec_keyed_map_array_container,
+                                                      is_array_container)
+
+        def to_arg_descr(arg: Any) -> AbstractInputDescriptor:
+            if np.isscalar(arg):
+                return ScalarInputDescriptor(np.dtype(arg))
+            elif is_array_container(arg):
+                id_to_ary_descr = {}
+
+                def id_collector(keys, ary):
+                    id_to_ary_descr[keys] = (np.dtype(ary.dtype),
+                                             ary.shape)
+                    return ary
+
+                rec_keyed_map_array_container(id_collector, arg)
+                return ArrayContainerInputDescriptor(pmap(id_to_ary_descr))
+            else:
+                raise ValueError("Argument to a compiled operator should be"
+                                 " either a scalar or an array container. Got"
+                                 f" '{arg}'.")
+
+        arg_descrs = tuple(to_arg_descr(arg) for arg in args)
+
+        try:
+            exec_f = self.program_cache[arg_descrs]
+        except KeyError:
+            pass
+        else:
+            return exec_f(*args)
+
+        dict_of_named_arrays = {}
+        # output_naming_map: result id to name of the named array in the
+        # generated pytato DAG.
+        output_naming_map = {}
+        # input_naming_map: argument id to placeholder name in the generated
+        # pytato DAG.
+        input_naming_map = {}
+
+        def to_placeholder(arg, pos):
+            if np.isscalar(arg):
+                name = f"_actx_in_{pos}"
+                input_naming_map[(pos, )] = name
+                return pt.make_placeholder((), np.dtype(arg), name)
+            elif is_array_container(arg):
+                def _rec_to_placeholder(keys, ary):
+                    name = (f"_actx_in_{pos}_"
+                            + _ary_container_key_stringifier(keys))
+                    input_naming_map[(pos,) + keys] = name
+                    return pt.make_placeholder(ary.shape, ary.dtype,
+                                               name)
+                return rec_keyed_map_array_container(_rec_to_placeholder,
+                                                     arg)
+            else:
+                raise NotImplementedError(type(arg))
+
+        outputs = self.f(*[to_placeholder(arg, iarg)
+                           for iarg, arg in enumerate(args)])
+
+        if not is_array_container(outputs):
+            # TODO: We could possibly just short-circuit this interface if the
+            # returned type is a scalar. Not sure if it's worth it though.
+            raise ValueError(f"Function '{self.f.__name__}' to be compiled did not"
```

Comment:
```suggestion
            raise NotImplementedError(f"Function '{self.f.__name__}' to be compiled did not"
```
(I think we'll want this before too long.)